aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,389 @@
+//! let privkey_raw_bytes = priv_key.get_raw_bytes();
+//!
+//! let pub_key = priv_key.compute_public_key()?;
+//!
+//! // Get the public key bytes to send to bob through some encoded
+//! // protocol message.
+//! let pub_key_bytes = pub_key.get_raw_bytes();
+//!
+//! let mut ciphertext: Vec<u8> = vec![];
+//!
+//! let retrieved_pub_key = KemPublicKey::from_raw_bytes(KemAlgorithm::KYBER512_R3, pub_key_bytes)?;
+//! let bob_result = retrieved_pub_key.encapsulate(|ct, ss| {
+//!     ciphertext.extend_from_slice(ct);
+//!     // In a real application, we'd apply a KDF to the shared secret and the
+//!     // public keys (as recommended in RFC 7748) and then derive session
+//!     // keys from the result. We omit all that here.
+//!     Ok(())
+//! });
+//! 
+//! // Retrieve private key from stored raw bytes
+//! let retrieved_priv_key = KemPrivateKey::from_raw_bytes(KemAlgorithm::KYBER512_R3, privkey_raw_bytes)?;
+//!
+//! let alice_result = retrieved_priv_key.decapsulate(&mut ciphertext, |ss| {
+//!     // In a real application, we'd apply a KDF to the shared secret and the
+//!     // public keys (as recommended in RFC 7748) and then derive session
+//!     // keys from the result. We omit all that here.
+//!     Ok(())
+//! });
+//! 
+//! # Ok::<(), aws_lc_rs::error::Unspecified>(())
+//! ```
+use crate::{error::{Unspecified, KeyRejected}, ptr::LcPtr, ptr::{DetachableLcPtr, Pointer}};
+use std::{os::raw::c_int};
+use std::ptr::null_mut;
+use aws_lc::{
+    EVP_PKEY, NID_KYBER512_R3, EVP_PKEY_keygen, EVP_PKEY_CTX_new_id, EVP_PKEY_keygen_init,
+    EVP_PKEY_KEM, EVP_PKEY_kem_new_raw_secret_key, EVP_PKEY_CTX_kem_set_params, EVP_PKEY_CTX_new,
+    EVP_PKEY_encapsulate, EVP_PKEY_decapsulate, EVP_PKEY_get_raw_public_key, EVP_PKEY_kem_new_raw_public_key, EVP_PKEY_get_raw_private_key
+};
+use zeroize::Zeroize;
+
+
+const KYBER512_SECRETKEYBYTES: usize = 1632;
+const KYBER512_PUBLICKEYBYTES: usize = 800;
+const KYBER512_CIPHERTEXTBYTES: usize = 768;
+const KYBER512_BYTES: usize = 32;
+
+
+#[allow(non_camel_case_types)]
+#[derive(Clone, Debug, PartialEq)]
+/// A selection of algorithms to be used as KEMs.
+pub enum KemAlgorithm {
+    /// NIST Round 3 iteration of the Kyber-512 algorithm
+    KYBER512_R3,
+}
+
+impl KemAlgorithm {
+    #[inline]
+    fn nid(&self) -> i32 {
+        match self {
+            KemAlgorithm::KYBER512_R3 => NID_KYBER512_R3,
+        }
+    }
+
+    #[inline]
+    fn get_pubkey_len(&self) -> usize {
+        match self {
+            KemAlgorithm::KYBER512_R3 => KYBER512_PUBLICKEYBYTES,
+        }
+    }
+
+    #[inline]
+    fn get_privkey_len(&self) -> usize {
+        match self {
+            KemAlgorithm::KYBER512_R3 => KYBER512_SECRETKEYBYTES,
+        }
+    }
+}
+/// A serializable private key usable with KEMs. This can be randomly generated
+/// or constructed from raw bytes.
+#[derive(Debug)]
+pub struct KemPrivateKey {
+    algorithm: KemAlgorithm,
+    context: LcPtr<*mut EVP_PKEY>,
+    priv_key: Vec<u8>
```

Comment:
It's necessary for us to do this if we want to return an `&[u8]` in the `get_raw_bytes` method. I think a Vec makes the implementation easier because we don't always know the length of what we are allocating at compile time. Why would a Box<[u8]> be better?